### PR TITLE
Aws alb schema

### PR DESCRIPTION
**Pull request by thewizarodofoz**
Closes: #8020 

Notes:
1. I Left `listenerArn.match(ALB_LISTENER_PATTERN)` in `validate.js` because it's used as a parser. Feels weird importing it in `awsProvider` for validating but couldn't think of a better way
2. `conditions.header.name` - I didn't specify the pattern, seems like an overkill
3. `healthCheck.mather.httpCode` - I didn't specify the pattern, found it very hard to do

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/alb/index.js [hunk 1]
```diff
@@ -28,8 +28,81 @@ class AwsCompileAlbEvents {
       },
     };
 
-    // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8020
-    this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'alb', { type: 'object' });
+    this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'alb', {
+      type: 'object',
+      properties: {
+        authorizer: { type: 'string' },
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/alb/index.js:34]: It's `string` or `array`

File: lib/plugins/aws/package/compile/events/alb/index.js [hunk 2]
```diff
@@ -28,8 +28,81 @@ class AwsCompileAlbEvents {
...
+            { type: 'boolean' },
+            {
+              type: 'object',
+              properties: {
+                healthyThresholdCount: { type: 'integer', minimum: 2, maximum: 10 },
+                intervalSeconds: { type: 'integer', minimum: 5, maximum: 300 },
+                matcher: {
+                  type: 'object',
+                  properties: {
+                    httpCode: { type: 'string' },
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/alb/index.js:86]: We can probably restrict this with `pattern` as `^\d{3}(-\d{3})?(,\d{3}(-\d{3})?)*$`

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-targetgroup-matcher.html

File: lib/plugins/aws/provider/awsProvider.js [hunk 3]
```diff
@@ -325,6 +330,74 @@ class AwsProvider {
...
+              type: 'object',
+              properties: {
+                targetGroupPrefix: { type: 'string', maxLength: 16 },
+                authorizers: {
+                  type: 'object',
+                  additionalProperties: {
+                    type: 'object',
+                    properties: {
+                      type: { enum: ['cognito', 'oidc'] },
+                      userPoolArn: { $ref: '#/definitions/awsArnString' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:343]: I think, even though documentation states _string_, the CloudFormation intrinsic functions are naturally supported. Therefore it should rather be `awsArn` and not `awsArnString`

File: lib/plugins/aws/provider/awsProvider.js [hunk 4]
```diff
@@ -325,6 +330,74 @@ class AwsProvider {
...
+                      tokenEndpoint: { format: 'uri' },
+                      userInfoEndpoint: { format: 'uri' },
+                      requestExtraParams: {
+                        type: 'object',
+                        maxProperties: 10,
+                        additionalProperties: { type: 'string' },
+                      },
+                      scope: { type: 'string' },
+                      sessionCookieName: { type: 'string' },
+                      sessionTimeout: { type: 'integer' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:361]: We may add `minimum: 0` (I imagine negative values not being supported)

File: lib/plugins/aws/provider/awsProvider.js [hunk 5]
```diff
@@ -325,6 +330,74 @@ class AwsProvider {
...
+                    properties: {
+                      type: { enum: ['cognito', 'oidc'] },
+                      userPoolArn: { $ref: '#/definitions/awsArnString' },
+                      userPoolClientId: { type: 'string' },
+                      userPoolDomain: { type: 'string' },
+                      onUnauthenticatedRequest: { enum: ['allow', 'authenticate', 'deny'] },
+                      authorizationEndpoint: { format: 'uri' },
+                      clientId: { type: 'string' },
+                      clientSecret: { type: 'string' },
+                      useExistingClientSecret: { type: 'boolean' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:350]: I think we do not support such property

File: lib/plugins/aws/provider/awsProvider.js [hunk 6]
```diff
@@ -325,6 +330,74 @@ class AwsProvider {
...
+                            'issuer',
+                            'tokenEndpoint',
+                            'userInfoEndpoint',
+                          ],
+                          oneOf: [
+                            { required: ['clientSecret'] },
+                            { required: ['useExistingClientSecret'] },
+                          ],
+                        },
+                      },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:393]: I think it'll be nicer to define both authorizers indepenedently, and put them into `oneOf`.

For those few properties which are shared we may put definitions into variables, so maintainance wise code is not duplicated (and it doesn't serve as good case to create new definition)

File: lib/plugins/aws/provider/awsProvider.js [hunk 7]
```diff
@@ -325,6 +330,74 @@ class AwsProvider {
...
+                    ],
+                    required: ['type'],
+                    additionalProperties: false,
+                  },
+                },
+              },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:399]: We miss `additionalProperties: false`

File: lib/plugins/aws/package/compile/events/alb/lib/validate.js [hunk 8]
```diff
@@ -58,13 +47,13 @@ module.exports = {
...
-              albObj.conditions.header = this.validateHeaderCondition(event, functionName);
+              albObj.conditions.header = event.alb.conditions.header;
             }
             if (event.alb.conditions.query) {
-              albObj.conditions.query = this.validateQueryCondition(event, functionName);
+              albObj.conditions.query = event.alb.conditions.query;
             }
             if (event.alb.conditions.ip) {
-              albObj.conditions.ip = this.validateIpCondition(event, functionName);
+              albObj.conditions.ip = event.alb.conditions.ip;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/alb/lib/validate.js:56]: We need to maintain support for `string` notation, so it should be `[].concat(event.alb.conditions.ip)`

File: lib/plugins/aws/package/compile/events/alb/lib/validate.js [hunk 9]
```diff
@@ -89,21 +78,13 @@ module.exports = {
...
     if (_.isObject(listenerArn)) {
       if (listenerArn.Ref) {
         return { albId: listenerArn.Ref, listenerId: listenerArn.Ref };
       }
-      throw new this.serverless.classes.Error(
-        `Invalid ALB listenerArn in function "${functionName}".`
-      );
     }
-    const matches = listenerArn.match(ALB_LISTENER_PATTERN);
+    const matches = listenerArn.match(new RegExp(this.ALB_LISTENER_PATTERN));
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/alb/lib/validate.js:87]: It'll be nicer to define `this.ALB_LISTENER_PATTERN` as regExp instance, and for config schema, simply coerce it to string via `String(this.ALB_LISTENER_PATTERN)`

With above we create a new object instance for one time validation and it doesn't feel clean.

**Review by medikoo — CHANGES_REQUESTED** [review]
wow, that looks great @thewizarodofoz ! I've spotted just some minor shortcomings, please see my comments

File: lib/plugins/aws/package/compile/events/alb/index.js [hunk 10]
```diff
@@ -28,8 +28,81 @@ class AwsCompileAlbEvents {
...
+              additionalProperties: false,
+            },
+          ],
+        },
+        listenerArn: {
+          oneOf: [{ $ref: '#/definitions/awsAlbListenerArn' }, { $ref: '#/definitions/awsCfRef' }],
+        },
+        priority: { type: 'integer', minimum: 1, maximum: 50000 },
+      },
+      required: ['listenerArn', 'priority'],
```
[the remark this answers is not in the document]
Comment (thewizarodofoz) [reply lib/plugins/aws/package/compile/events/alb/index.js:103]: According to [this line](url) we might even say `conditions.path` is required. WDYT @medikoo ?

File: lib/plugins/aws/package/compile/events/alb/lib/validate.js (hunk 9 above)
Comment (thewizarodofoz) [reply lib/plugins/aws/package/compile/events/alb/lib/validate.js:87]: even `.toString()` works

File: lib/plugins/aws/package/compile/events/alb/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/alb/index.js:103]: Yes, I believe we should also make it required

File: lib/plugins/aws/package/compile/events/alb/lib/validate.js (hunk 9 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/alb/lib/validate.js:87]: Yes, it'll return regexp pattern as string (`String(regExp)` and `regExp.toString()` are equivalent)

Comment (thewizarodofoz) [reply lib/plugins/aws/package/compile/events/alb/lib/validate.js:87]: actually both `String(regExp)` and `regExp.toString()` output the enclosing forward slashes, but ajv expects patterns without them. `regExp.source` is what we want.

Comment (medikoo) [reply lib/plugins/aws/package/compile/events/alb/lib/validate.js:87]: Indeed, I forgot about that, yes `regExp.source` is the way to go

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks @thewizarodofoz for update! Still, I see some comments not addressed/answered, have you pushed all the commits?

**Review by medikoo — CHANGES_REQUESTED** [review]
Thanks @thewizarodofoz, still as I see CI tests fail because regex is not right for `httpCode` (It has literal `d` instead of `\d` which defines digit - in a string we need to input as `\\d`)

File: lib/plugins/aws/provider/awsProvider.js [hunk 11]
```diff
@@ -123,6 +124,54 @@ const impl = {
...
+    'userInfoEndpoint',
+  ],
+  additionalProperties: false,
+};
+
+const cognitoAlbAuthorizer = {
+  type: 'object',
+  properties: {
+    type: { const: 'cognito' },
+    userPoolArn: { $ref: '#/definitions/awsArnString' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:166]: This seems not addressed https://github.com/serverless/serverless/pull/8291#discussion_r495842300

File: lib/plugins/aws/provider/awsProvider.js [hunk 12]
```diff
@@ -123,6 +124,54 @@ const impl = {
...
+const baseAlbAuthorizerProperties = {
+  onUnauthenticatedRequest: { enum: ['allow', 'authenticate', 'deny'] },
+  requestExtraParams: {
+    type: 'object',
+    maxProperties: 10,
+    additionalProperties: { type: 'string' },
+  },
+  scope: { type: 'string' },
+  sessionCookieName: { type: 'string' },
+  sessionTimeout: { type: 'integer' },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:136]: This seems not addressed https://github.com/serverless/serverless/pull/8291#discussion_r495843839

**Review by medikoo — COMMENTED** [review]
@thewizarodofoz looks great! I've just spotted two suggestions not answered/addressed. See my comments

File: lib/plugins/aws/provider/awsProvider.js (hunk 11 above)
Comment (thewizarodofoz) [reply lib/plugins/aws/provider/awsProvider.js:166]: Sorry about all this back&forth, I keep forgetting to unfold the hidden comments so I keep missing a few.

**Review by medikoo — APPROVED** [review]
Thank you @thewizarodofoz !

**Pull request merged**